File: rs/src/protocol.rs
pub mod auth {
    include!(concat!(env!("OUT_DIR"), "/spire.protocol.auth.rs"));
}

pub mod game {
    include!(concat!(env!("OUT_DIR"), "/spire.protocol.game.rs"));
}

pub mod net {
    include!(concat!(env!("OUT_DIR"), "/spire.protocol.net.rs"));
}

pub enum Protocol {
    None = 0,
    Auth = 1,
    Game = 2,
    Net = 3,
}

pub fn write_header(protocol: Protocol, length: u16, buf: &mut [u8; 4]) {
    const RESERVED: u8 = 0u8;

    buf[0] = protocol as u8;
    buf[1] = RESERVED;
    buf[2] = (length >> 8) as u8;
    buf[3] = length as u8;
}

pub fn read_header(buf: &[u8; 4]) -> (Protocol, u16) {
    let protocol = match buf[0] {
        1 => Protocol::Auth,
        2 => Protocol::Game,
        3 => Protocol::Net,
        _ => Protocol::None,
    };
    let length =  ((buf[2] as u16) << 8) | (buf[3] as u16);

    (protocol, length)
}

File: cs/src/Protocol.cs
using System;

namespace spire.protocol;

public enum ProtocolCategory : byte
{
    None = 0,
    Auth = 1,
    Game = 2,
    Net = 3
}

public static class ProtocolHeader
{
    public const uint HeaderSize = 4;
    private const byte Reserved = 0;

    public static void Write(ProtocolCategory protocolCategory, ushort length, Span<byte> buffer)
    {
        if (buffer.Length < 4)
            throw new ArgumentOutOfRangeException("Buffer must be at least 4 bytes long.");

        buffer[0] = (byte)protocolCategory;
        buffer[1] = Reserved;
        buffer[2] = (byte)(length >> 8);
        buffer[3] = (byte)length;
    }

    public static (ProtocolCategory, ushort) Read(ReadOnlySpan<byte> buffer)
    {
        if (buffer.Length < 4)
            throw new ArgumentOutOfRangeException("Buffer must be at least 4 bytes long.");

        var protocol = buffer[0] switch
        {
            1 => ProtocolCategory.Auth,
            2 => ProtocolCategory.Game,
            3 => ProtocolCategory.Net,
            _ => ProtocolCategory.None
        };
        var length = (ushort)((buffer[2] << 8) | buffer[3]);

        return (protocol, length);
    }
}